File: C++/OOP Lab4/mushrooms.cpp
#include "mushrooms.h"

Mushroom::Mushroom(std::string name) : name(name) {};
std::string Mushroom::getName() {
	return name;
}
std::ostream& operator << (std::ostream& out, Mushroom& mushroom) {
	return mushroom.output(out, mushroom);
}

EatableMushroom::EatableMushroom(std::string name) : Mushroom(name), eatable(true) {};
std::ostream& EatableMushroom::output(std::ostream& out, Mushroom& mushroom) {
	return out << mushroom.getName() << ", " << mushroom.isEatable() << ", 0 minutes";
}
bool EatableMushroom::isEatable() {
	return eatable;
}
int EatableMushroom::getTreatment() {
	return 0;
}


PoisonousMushroom::PoisonousMushroom(std::string name, int treatment) : Mushroom(name), eatable(false), treatment(treatment) {};
std::ostream& PoisonousMushroom::output(std::ostream& out, Mushroom& mushroom) {
	return out << mushroom.getName() << ", " << mushroom.isEatable() << ", " << mushroom.getTreatment() << " minutes";
}
int PoisonousMushroom::getTreatment() {
	return treatment;
}
bool PoisonousMushroom::isEatable() {
	return eatable;
}

MushroomContext::MushroomContext(std::string nname, bool eatable, int ttreatment) : name(nname), treatment(ttreatment) {
	if (eatable) mushroomType = Eatable; else mushroomType = Poisonous;
}
MushroomType MushroomContext::getMushroomType() {
	return mushroomType;
}
std::string MushroomContext::getName() {
	return name;
}
int MushroomContext::getTreatment() {
	return treatment;
}

Mushroom* MushroomFactory::createMushroom(MushroomContext context) {
	switch (context.getMushroomType())
	{
	case Eatable:
		return new EatableMushroom(context.getName());
	case Poisonous:
		return new PoisonousMushroom(context.getName(), context.getTreatment());
	default:
		std::cerr << "Unknown mushroom type: " << context.getMushroomType() << "!" << std::endl;
		system("pause");
		exit(0);
	}
}

File: Python/mushrooms.py
from abc import ABC, abstractmethod
from enum import Enum, auto


class MushroomType(Enum):
    POISONOUS = auto()
    EATABLE = auto()


class Mushroom(ABC):
    def __init__(self, name):
        self._name = name

    @abstractmethod
    def get_name(self) -> str:
        pass


class EatableMushroom(Mushroom):
    def __init__(self, name):
        super().__init__(name)
        self._eatable = True

    def __str__(self):
        return f"{self._name}, true, 0 minutes"

    def get_name(self):
        return self._name

    def is_eatable(self):
        return self._eatable


class PoisonousMushroom(Mushroom):
    def __init__(self, name, treatment: int):
        super().__init__(name)
        self._eatable = False
        self._treatment = treatment

    def __str__(self):
        return f"{self._name}, false, {self._treatment} minutes"

    def get_name(self):
        return self._name

    def get_treatment(self):
        return self._treatment

    def is_eatable(self):
        return self._eatable


class MushroomFactory:
    @staticmethod
    def create_mushroom(context):
        if context.mushroom_type == MushroomType.EATABLE:
            return EatableMushroom(context.name)
        elif context.mushroom_type == MushroomType.POISONOUS:
            return PoisonousMushroom(context.name, context.treatment)
        else:
            raise ValueError(f"Unknown mushroom type: {context.mushroom_type}!")


class MushroomContext:
    def __init__(self, name, eatable, treatment: int):
        self.mushroom_type = MushroomType.EATABLE if eatable else MushroomType.POISONOUS
        self.name = name
        self.treatment = treatment
